Upstream checks the .note.ABI-tag in a bad way.
They should not only check the "GNU", but also whether OS field matches.
In both cases they should check 8 bytes, we can use it for current hack.

http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=532627
http://www.freebsd.org/cgi/query-pr.cgi?pr=135468


Index: src/sys/amd64/linux32/linux32_sysvec.c
===================================================================
--- src.orig/sys/amd64/linux32/linux32_sysvec.c
+++ src/sys/amd64/linux32/linux32_sysvec.c
@@ -1047,10 +1047,10 @@
 	.sv_maxssiz	= &linux32_maxssiz,
 };
 
-static char GNULINUX_ABI_VENDOR[] = "GNU";
+static char GNULINUX_ABI_VENDOR[] = "GNU\0\0\0\0\0";
 
 static Elf_Brandnote linux32_brandnote = {
-	.hdr.n_namesz	= sizeof(GNULINUX_ABI_VENDOR),
+	.hdr.n_namesz	= 4,
 	.hdr.n_descsz	= 16,
 	.hdr.n_type	= 1,
 	.vendor		= GNULINUX_ABI_VENDOR,
Index: src/sys/i386/linux/linux_sysvec.c
===================================================================
--- src.orig/sys/i386/linux/linux_sysvec.c
+++ src/sys/i386/linux/linux_sysvec.c
@@ -1017,10 +1017,10 @@
 	.sv_maxssiz	= NULL
 };
 
-static char GNULINUX_ABI_VENDOR[] = "GNU";
+static char GNULINUX_ABI_VENDOR[] = "GNU\0\0\0\0\0";
 
 static Elf_Brandnote linux_brandnote = {
-	.hdr.n_namesz	= sizeof(GNULINUX_ABI_VENDOR),
+	.hdr.n_namesz	= 4,
 	.hdr.n_descsz	= 16,
 	.hdr.n_type	= 1,
 	.vendor		= GNULINUX_ABI_VENDOR,
Index: src/sys/kern/imgact_elf.c
===================================================================
--- src.orig/sys/kern/imgact_elf.c
+++ src/sys/kern/imgact_elf.c
@@ -112,10 +112,10 @@
 #define	round_page_ps(va, ps)	(((va) + (ps - 1)) & ~(ps - 1))
 #define	aligned(a, t)	(trunc_page_ps((u_long)(a), sizeof(t)) == (u_long)(a))
 
-static const char FREEBSD_ABI_VENDOR[] = "FreeBSD";
+static const char FREEBSD_ABI_VENDOR[] = "FreeBSD\0";
 
 Elf_Brandnote __elfN(freebsd_brandnote) = {
-	.hdr.n_namesz	= sizeof(FREEBSD_ABI_VENDOR),
+	.hdr.n_namesz	= 8,
 	.hdr.n_descsz	= sizeof(int32_t),
 	.hdr.n_type	= 1,
 	.vendor		= FREEBSD_ABI_VENDOR,
@@ -1393,8 +1393,7 @@
 		    note->n_type != checknote->hdr.n_type)
 			goto nextnote;
 		note_name = (const char *)(note + 1);
-		if (strncmp(checknote->vendor, note_name,
-		    checknote->hdr.n_namesz) != 0)
+		if (memcmp(checknote->vendor, note_name, 8) != 0)
 			goto nextnote;
 
 		/*
